### PR TITLE
upgrades: use version gate to see if role members table has id columns

### PR DESCRIPTION
**Pull request by healthy-pod**
This code change makes the `addRootUser` permanent upgrade
depend solely on the version gate to check if
`V23_1RoleMembersTableHasIDColumns` migration ran.

It also updates tests that were failing because they
bootstrap with unexpected versions to bootstrap with
the right versions.

Release note: None
Epic: CRDB-21741

### CODE REVIEW

File: pkg/upgrade/upgrades/permanent_upgrades.go [hunk 1]
```diff
@@ -50,39 +49,23 @@ func addRootUser(
...
-	// We intentionally override any existing entry.
-	return deps.DB.Txn(ctx, addRootToAdminRole)
-}
-
-func addRootToAdminRole(ctx context.Context, txn isql.Txn) error {
 	var upsertStmt string
 	var upsertVals []interface{}
-	{
-		// We query the pg_attribute to determine whether the role_id and member_id
-		// columns are present because we can't rely on version gates here.
```
Comment (andyyang890) [inline pkg/upgrade/upgrades/permanent_upgrades.go:64]: Did you determine that this comment was/is no longer true? I did take a look at #94774 (the PR where it was added) just now and I'm not quite sure what the initial reasoning for it was.

Comment (healthy-pod) [reply pkg/upgrade/upgrades/permanent_upgrades.go:64]: I think this comment is not true. Tests failed because they were passing version overrides that are not compatible with the actual schema / initial values used.

They had an active version that is lower than the version gate (due to a version override) but in reality they ran the migration that adds these columns and hence need to pass a non-null value.

In the future, tests passing version overrides should ensure that they also pass an appropriate value for `BootstrapVersionKeyOverride`.

File: pkg/server/version_cluster_test.go [hunk 2]
```diff
@@ -414,6 +414,7 @@ func TestClusterVersionMixedVersionTooOld(t *testing.T) {
 	knobs := base.TestingKnobs{
 		Server: &server.TestingKnobs{
 			DisableAutomaticVersionUpgrade: make(chan struct{}),
+			BootstrapVersionKeyOverride:    clusterversion.BinaryMinSupportedVersionKey,
```
Comment (andyyang890) [inline pkg/server/version_cluster_test.go:417]: Is it okay that the binary version is going to be lower than the bootstrap version? From looking at the definition of `v0v1` above, it seems that `v0` is going to be one version lower than `clusterversion.BinaryMinSupportedVersionKey`. I guess ideally it would use `v0` as the bootstrap version but it seems like that version might either not exist (the code in `v0v1` seems to allow `0` as the minor version) or be deleted already.

File: pkg/kv/kvserver/client_lease_test.go [hunk 3]
```diff
@@ -1537,79 +1536,6 @@ func TestLeaseTransfersUseExpirationLeasesAndBumpToEpochBasedOnes(t *testing.T)
 	require.Equal(t, roachpb.LeaseExpiration, mu.lease.Type())
 }
 
-// TestLeaseUpgradeVersionGate tests the version gating for the lease-upgrade
```
Comment (andyyang890) [inline pkg/kv/kvserver/client_lease_test.go:1540]: Why are we deleting this test? Also, can you move the delete into a separate commit?

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go [hunk 4]
```diff
@@ -260,71 +257,6 @@ func channelWaitWithTimeout(t *testing.T, ch chan struct{}) {
 		t.Fatal("test timed out")
 	}
 }
-func TestBiDirectionalRangefeedNotUsedUntilUpgradeFinalilzed(t *testing.T) {
```
Comment (andyyang890) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go:263]: Why are we deleting this test? Also, can you move the delete into a separate commit?

File: pkg/server/version_cluster_test.go (hunk 2 above)
Comment (healthy-pod) [reply pkg/server/version_cluster_test.go:403]: Simplified the test to use `TestingBinaryMinSupportedVersion` and `TestingBinaryVersion ` instead.

File: pkg/kv/kvserver/client_lease_test.go (hunk 3 above)
Comment (healthy-pod) [reply pkg/kv/kvserver/client_lease_test.go]: Sure, deleted in #111396

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go (hunk 4 above)
Comment (healthy-pod) [reply pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go]: Sure, deleted in #111396

Comment (healthy-pod) [reply pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go]: This and the other test were deleted because they are no longer needed in the current branch (they were testing version gates that no longer exist).

**Comment by stevendanna** [comment]
Do we have any worries about the kind of issues we've encountered with other version-gate based queries against system tables: https://github.com/cockroachdb/cockroach/issues/109039

**Comment by stevendanna** [comment]
I defer to SQL foundations on this one, so please don't block on me.

**Comment by healthy-pod** [comment]
> Do we have any worries about the kind of issues we've encountered with other version-gate based queries against system tables: https://github.com/cockroachdb/cockroach/issues/109039

Thanks for pointing this out, I forgot about the `AS OF SYSTEM TIME` issue.

I tried reworking this to use `GetSystemDatabaseSchemaVersion` from Andy's [PR](https://github.com/cockroachdb/cockroach/pull/111404) but unfortunately that won't work with tests that boostrap using older versions.

We should keep making the check the same way for now. I will create an issue for SQL Foundations to move to using `GetSystemDatabaseSchemaVersion` when the minimum supported binary version aka the minimum version tests can bootstrap at is `23.2`.

**Comment by healthy-pod** [comment]
Opened https://github.com/cockroachdb/cockroach/issues/111894 and will close this PR.
